[Android] Fix for single char name for Wii and WAD Games

34will opened the pull request:
The issue seems to be when inserting the value into the GameDatabase. The value is correctly read from the file, but every character seems to be alternated with a null-terminating character, `\0`, thus when the string is returned to the Java environment via `.c_str()`, only the first character is then shown. I added a RemoveNullTermChars method to strip them out, and applied it to the GetTitle, GetDescription, and GetCompany methods.

JosJuice commented:
I'm not sure if just removing all null chars is the correct fix. This kind of thing usually happens when something is in UTF-16 but the code doesn't treat it as UTF-16 (not that I know how that would be happening in this case). Have you tested this with a game that uses non-ASCII characters, such as a Pokémon game or a Japanese game?

JosJuice commented:
If Wii games give you these null bytes but not GameCube games, my guess is that the following function (or the `UTF16ToUTF8` function that it calls) is broken on Android:

https://github.com/dolphin-emu/dolphin/blob/835eff15a18fab15ee8c08241dd9b1588dd428ca/Source/Core/Common/StringUtil.cpp#L581-L587

It's supposed to return a string that doesn't contain any null bytes.

JosJuice commented:
Correction to my first comment: I forgot that Android Dolphin tries to show the English name for every game no matter what, even if it's a Japanese game... So if you want to do some testing with Japanese strings, replace the instances of `DiscIO::Language::LANGUAGE_ENGLISH` in MainAndroid.cpp with `DiscIO::Language::LANGUAGE_JAPANESE`. This will give you Japanese strings for many Wii games, even ones that you don't actually have a Japanese copy of.

34will commented:
Okay, I have done some digging on the issue. It seems that the problem goes further than `UTF16BEToUTF8()` or `UTF16ToUTF8()`, the call to the libiconv library itself: https://github.com/dolphin-emu/dolphin/blob/835eff15a18fab15ee8c08241dd9b1588dd428ca/Source/Core/Common/StringUtil.cpp#L515-L516 returns the text with the null characters in it, so it seems to be an issue with Android, as correctly ascertained by @JosJuice. 

Digging further found [this stackoverflow](https://stackoverflow.com/questions/8265776/how-to-convert-wchar-wstring-in-c-to-something-better-supported), which references [this article](http://www.charleswilde.com/porting-applications-using-the-android-ndk/), which seems to suggest that Android stores `wchar_t`s as 4 bytes as opposed to the 2 bytes required for UTF-16. So when the data eventually gets to the libiconv call, it's expecting 2 bytes per character, but is getting 4-per, so it treats each actual character as 2 characters, the correct one and an empty one, which it resolves to a null character, leading to the alternating nulls. The article mentions a flag that can be passed to the C++ compiler to force it to store `wchar_t`s as 2 bytes, which I tried, and whilst it did succeed in changing the size of the characters, the app started throwing `SIGABRT` and `SIGSEGV` errors, so it wasn't really viable.

However as I was looking at the code, I saw that there was no requirement to use a `wstring`, so I modified it to use a `std::u16string` which is 2 bytes per character, standard guaranteed, and the libiconv code seemed able to handle it, and it produced the correct output first time. I also tested it with Japanese characters and they worked fine too.

JosJuice commented on Source/Core/Common/StringUtil.cpp:
```diff
@@ -559,12 +567,12 @@ std::string UTF16ToUTF8(const std::wstring& input)
+  std::u16string result(static_cast<size_t>(str_end - str), '\0');
   std::transform(str, str_end, result.begin(), static_cast<u16 (&)(u16)>(Common::swap16));
-  return UTF16ToUTF8(result);
+  return CodeToUTF8("UTF-16LE", result);
```
I think it should be possible to remove the `std::transform` line if you replace `"UTF-16LE"` with `"UTF-16BE"`.

34will replied:
See commit a32d38817f6000594a757de683f5835b01073387.

34will commented:
I did as suggested and removed the extraneous things from `UTF16BEToUTF8` and removed the non-Windows `UTF16ToUTF8` method, and it works correctly on Android, but I don't have a Linux box to test it on.

JosJuice commented:
In the header file, you need to move `std::string UTF16ToUTF8(const std::wstring& str);` so that it's between `#ifdef _WIN32` and the corresponding `#endif`. (It probably fits best right next to `UTF8ToUTF16`.) Otherwise there might be unexpected linking errors.

34will commented:
Yep, just caught that. I'm always forgetting something with C++

JosJuice commented on Source/Core/Common/StringUtil.cpp:
```diff
@@ -554,17 +562,9 @@ std::string UTF8ToSHIFTJIS(const std::string& input)
+std::string UTF16BEToUTF8(const char16_t* str, size_t max_size)
 {
-  return CodeToUTF8("UTF-16LE", input);
+  return CodeToUTF8("UTF-16BE", std::u16string(str, max_size));
```
Doesn't this `u16string` constructor always read `max_size` code units? The number of code units that this function is supposed to read is whichever is smallest out of `max_size` and the position of the first null code unit. So I think you still need the line `const char16_t* str_end = std::find(str, str + max_size, '\0');`.

JosJuice replied:
It should then should be possible to construct the string using `std::u16string(str, str_end)`, but please correct me if I'm wrong.

34will replied:
Correct again! Sorry for the hand-holding.

JosJuice replied:
Don't worry about it. I'm just happy that someone cares enough about this issue to get it fixed :)

JosJuice commented:
Oops, seems like there actually was one place in the code that was using `UTF16ToUTF8` on platforms other than Windows after all... :/ I had missed that one.

Maybe we could write a new non-Windows implementation of `UTF16ToUTF8` that copies over all the characters to an `std::u16string` and then runs `CodeToUTF8` with `UTF-16LE`? Let's keep the new implementation of `UTF16BEToUTF8` that we already have in this PR, though.

34will commented:
Okay, I have done as requested, although 'copies over all the characters to an `std::u16string`' is not as easy as it looks. After a few hours of trawling the C++ Reference for UTF-16 conversions and the like, I came across another [stackoverflow answer](https://stackoverflow.com/questions/42734715/how-can-i-convert-wstring-to-u16string/42743775#42743775) which provides a rather manual method to convert a `srd::wstring` to a `std::u16string` and I shamelessly stole the code (I don't know if you guys have any kind of policy for that). I tested it on Android by putting the `UTF16BEToUTF8` method to back to how it used to work, by passing through to `UTF16ToUTF8`, and it seems to work well, in both English and Japanese.

JosJuice commented:
Ah, that code is doing a full-on conversion from UTF-32 to UTF-16... I didn't consider that that could be necessary when `wchar_t` is 32-bit. Yeah, the code does get pretty complicated then.

It should be possible to avoid needing `WStringToUTF16` if we implement `UTF16ToUTF8` using `codecvt_utf8<wchar_t>`: http://en.cppreference.com/w/cpp/locale/codecvt_utf8

34will commented:
That seems a bit simpler.

JosJuice commented:
@dolphin-emu-bot rebuild

degasus commented:
Sounds fine
